tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	if ((ret = tfw_http_field_val(hm, &s_field, &ua_value)) <= 0) {
+		return ret;
+	}
+
+	getnstimeofday(&ts);
+	nsecs = 1000ULL * 1000 * 1000 * ts.tv_sec + ts.tv_nsec;
+
+//	client->sock->sk_socket
+
+	crypto_shash_init(shash_desc);
+	crypto_shash_update(shash_desc, (u8 *)ua_value.ptr, ua_value.len);
+	crypto_shash_finup(shash_desc, (u8 *)&nsecs, sizeof(nsecs), client->hmac);
+
+	return 0;
+}
+
+/*
+ * Add Tempesta sticky cookie to an HTTP response.
+ *
+ * Create a complete 'Set-Cookie:' header field, and add it
+ * to the HTTP response' header block.
+ */
+static int
+tfw_http_sticky_add(TfwHttpMsg *hm, u_char *value, size_t len)
+{
+	int ret;
+	char buf[256], *ptr = buf;
+
+	/* Set up complete 'Set-Cookie:' header field */
+	memcpy(ptr, S_F_SET_COOKIE, SLEN(S_F_SET_COOKIE));
+	ptr += SLEN(S_F_SET_COOKIE);
+	memcpy(ptr, tfw_cfg_sticky.name.ptr, tfw_cfg_sticky.name.len);
+	ptr += tfw_cfg_sticky.name.len;
+	*ptr++ = '=';
+	ptr += tfw_http_prep_sticky(ptr, value, len);
+	TFW_DBG("%s: \"%.*s\"\n", __FUNCTION__, (int)(ptr - buf), buf);
+
+	if ((ret = tfw_http_hdr_add(hm, buf, ptr - buf)) != 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * No Tempesta sticky cookie found.
+ *
+ * Create Tempesta sticky cookie, and store it for future use.
+ * If configured, enforce Tempesta sticky cookie presence in all requests.
+ */
+static int
+tfw_http_sticky_notfound(TfwHttpMsg *hm)
+{
+	/* Create Tempesta sticky cookie and store it */
+	if (tfw_http_sticky_set(hm) != 0) {
+		return tfw_http_send_502(hm);
+	}
+	/*
+	 * If configured, make sure that backend server receives requests
+	 * that always have Tempesta sticky cookie. Return an HTTP 302
+	 * response to the client that has the same host, URI, and inludes
+	 * 'Set-Cookie' header field. Otherwise, forward the request to
+	 * a backend server.
+	 */
+	if (tfw_cfg_sticky.enforce) {
+		return tfw_http_send_302(hm);
+	}
+
+	return 0;
+}
+
+/*
+ * Found Tempesta sticky cookie.
+ */
+static int
+tfw_http_sticky_found(TfwHttpMsg *hm, tfw_str_t *value)
+{
+	TfwClient *client = (TfwClient *)hm->conn->peer;
+
+	/*
+	 * Do nothing for now. The request is passed to a backend server.
+	 */
+	TFW_DBG("Sticky cookie found: \"%.*s\"\n", (int)value->len, value->ptr);
+
+	return 0;
+}
+
+/*
+ * Process Tempesta sticky cookie in an HTTP request.
+ */
+int
+tfw_http_sticky_req_process(TfwHttpMsg *hm)
+{
+	int ret;
+	tfw_str_t value = { 0 };
+
+	if (tfw_cfg_sticky.enabled == 0) {
+		return 0;
+	}
+	/*
+	 * See if the Tempesta sticky cookie is present in the request,
+	 * and act depending on the result.
+	 */
+	ret = tfw_http_sticky_get(hm, &value);
+	if (ret < 0) {
+		return ret;
+	} else if (ret == 0) {
+		return tfw_http_sticky_notfound(hm);
+	} else if (ret == 1) {
+		return tfw_http_sticky_found(hm, &value);
+	} else {
+		TFW_WARN("Multiple Tempesta sticky cookies found: %d\n", ret);
+		return -1;
+	}
+	/*NOTREACHED*/
+	return -1;
```

Comment:
These 6 lines above should be just

         TFW_WARN("Multiple Tempesta sticky cookies found: %d\n", ret);
         return -1;